ch32v003fun/ch32v003fun.h

```diff
@@ -13794,7 +13794,29 @@ void SystemInit(void);
 // Put an output debug UART on Pin D5.
 // You can write to this with printf(...) or puts(...)
 
-#include <setjmp.h>
+#ifdef __riscv
+// _JBTYPE using long long to make sure the alignment is align to 8 byte,
+// otherwise in rv32imafd, store/restore FPR may mis-align.
+#define _JBTYPE long long
+#ifdef __riscv_e
+#define _JBLEN ((4*sizeof(long))/sizeof(long))
+#elif defined(__riscv_d)
+#define _JBLEN ((14*sizeof(long) + 12*sizeof(double))/sizeof(long))
+#elif defined(__riscv_f)
+#define _JBLEN ((14*sizeof(long) + 12*sizeof(float))/sizeof(long))
+#else
+#define _JBLEN ((14*sizeof(long))/sizeof(long))
+#endif
+#endif
```

Comment:
If this works then ir's probably fine but based on the comments here (https://github.com/riscv-collab/riscv-gnu-toolchain/issues/1555) the relevant checks should probably be done based on the ABI macro rather that other macros?

Reply:
Would you rather me use these?
```c
#define __riscv_float_abi_double 1 // ilp32d
#define __riscv_float_abi_single 1 // ilp32f
#define __riscv_float_abi_soft 1   // ilp32
#define __riscv_abi_rve 1          // ilp32e
```
Would be nice if GCC actually wrote down their define logic...

Reply:
> Would you rather me use these?
> 
> ```c
> #define __riscv_float_abi_double 1 // ilp32d
> #define __riscv_float_abi_single 1 // ilp32f
> #define __riscv_float_abi_soft 1   // ilp32
> #define __riscv_abi_rve 1          // ilp32e
> ```
> 
> Would be nice if GCC actually wrote down their define logic...

I presume you're referring to these? 

- https://github.com/riscv-non-isa/riscv-c-api-doc/blob/main/src/c-api.adoc#abi-related-preprocessor-definitions

Reply:
thank you!